File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/IeeeC37_118/ConfigurationCell.vb
'*******************************************************************************************************
'  ConfigurationCell.vb - IEEE C37.118 Configuration cell
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  11/12/2004 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.Text
Imports Tva.Phasors.Common
Imports Tva.Phasors.IeeeC37_118.Common

Namespace IeeeC37_118

    <CLSCompliant(False), Serializable()> _
    Public Class ConfigurationCell

        Inherits ConfigurationCellBase

        ' Because the protocol doesn't include a version number that can account draft implementations, we must manually account for this where needed
        Private m_formatFlags As FormatFlags
        Private m_configurationCount As UInt16

        Protected Sub New()
        End Sub

        Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

            MyBase.New(info, context)

            ' Deserialize configuration cell
            m_formatFlags = info.GetValue("formatFlags", GetType(FormatFlags))
            m_configurationCount = info.GetUInt16("configurationCount")

        End Sub

        Public Sub New(ByVal parent As ConfigurationFrame, ByVal idCode As UInt16, ByVal nominalFrequency As LineFrequency)

            MyBase.New(parent, False, idCode, nominalFrequency, MaximumPhasorValues, MaximumAnalogValues, MaximumDigitalValues)

        End Sub

        Public Sub New(ByVal configurationCell As IConfigurationCell)

            MyBase.New(configurationCell)

        End Sub

        ' This constructor satisfies ChannelCellBase class requirement:
        '   Final dervived classes must expose Public Sub New(ByVal parent As IChannelFrame, ByVal state As IChannelFrameParsingState, ByVal index As Int32, ByVal binaryImage As Byte(), ByVal startIndex As Int32)
        Public Sub New(ByVal parent As IConfigurationFrame, ByVal state As IConfigurationFrameParsingState, ByVal index As Int32, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

            ' We pass in defaults for id code and nominal frequency since these will be parsed out later
            MyBase.New(parent, False, MaximumPhasorValues, MaximumAnalogValues, MaximumDigitalValues, _
                New ConfigurationCellParsingState( _
                    AddressOf IeeeC37_118.PhasorDefinition.CreateNewPhasorDefintion, _
                    AddressOf IeeeC37_118.FrequencyDefinition.CreateNewFrequencyDefintion, _
                    AddressOf IeeeC37_118.AnalogDefinition.CreateNewAnalogDefintion, _
                    AddressOf IeeeC37_118.DigitalDefinition.CreateNewDigitalDefintion), _
                binaryImage, startIndex)

        End Sub

        Friend Shared Function CreateNewConfigurationCell(ByVal parent As IChannelFrame, ByVal state As IChannelFrameParsingState(Of IConfigurationCell), ByVal index As Int32, ByVal binaryImage As Byte(), ByVal startIndex As Int32) As IConfigurationCell

            Return New ConfigurationCell(parent, state, index, binaryImage, startIndex)

        End Function

        ' TODO: May want to shadow all parents in final derived classes - also go through code and make sure all MustInherit class properties are overridable
        Public Shadows ReadOnly Property Parent() As ConfigurationFrame
            Get
                Return MyBase.Parent
            End Get
        End Property

        Public Overrides ReadOnly Property InheritedType() As System.Type
            Get
                Return Me.GetType
            End Get
        End Property

        Public Property FormatFlags() As FormatFlags
            Get
                Return m_formatFlags
            End Get
            Set(ByVal value As FormatFlags)
                m_formatFlags = value
            End Set
        End Property

        Public Property ConfigurationCount() As UInt16
            Get
                Return m_configurationCount
            End Get
            Set(ByVal value As UInt16)
                m_configurationCount = value
            End Set
        End Property

        Public Overrides Property PhasorDataFormat() As DataFormat
            Get
                Return IIf((m_formatFlags And FormatFlags.Phasors) > 0, DataFormat.FloatingPoint, DataFormat.FixedInteger)
            End Get
            Set(ByVal value As DataFormat)
                If value = DataFormat.FloatingPoint Then
                    m_formatFlags = m_formatFlags Or FormatFlags.Phasors
                Else
                    m_formatFlags = m_formatFlags And Not FormatFlags.Phasors
                End If
            End Set
        End Property

        Public Overrides Property PhasorCoordinateFormat() As CoordinateFormat
            Get
                Return IIf((m_formatFlags And FormatFlags.Coordinates) > 0, CoordinateFormat.Polar, CoordinateFormat.Rectangular)
            End Get
            Set(ByVal value As CoordinateFormat)
                If value = CoordinateFormat.Polar Then
                    m_formatFlags = m_formatFlags Or FormatFlags.Coordinates
                Else
                    m_formatFlags = m_formatFlags And Not FormatFlags.Coordinates
                End If
            End Set
        End Property

        Public Overrides Property FrequencyDataFormat() As DataFormat
            Get
                Return IIf((m_formatFlags And FormatFlags.Frequency) > 0, DataFormat.FloatingPoint, DataFormat.FixedInteger)
            End Get
            Set(ByVal value As DataFormat)
                If value = DataFormat.FloatingPoint Then
                    m_formatFlags = m_formatFlags Or FormatFlags.Frequency
                Else
                    m_formatFlags = m_formatFlags And Not FormatFlags.Frequency
                End If
            End Set
        End Property

        Public Overrides Property AnalogDataFormat() As DataFormat
            Get
                Return IIf((m_formatFlags And FormatFlags.Analog) > 0, DataFormat.FloatingPoint, DataFormat.FixedInteger)
            End Get
            Set(ByVal value As DataFormat)
                If value = DataFormat.FloatingPoint Then
                    m_formatFlags = m_formatFlags Or FormatFlags.Analog
                Else
                    m_formatFlags = m_formatFlags And Not FormatFlags.Analog
                End If
            End Set
        End Property

        Protected Overrides ReadOnly Property HeaderLength() As UInt16
            Get
                Return MyBase.HeaderLength + 10
            End Get
        End Property

        Protected Overrides ReadOnly Property HeaderImage() As Byte()
            Get
                Dim buffer As Byte() = CreateArray(Of Byte)(HeaderLength)
                Dim index As Int32

                CopyImage(MyBase.HeaderImage, buffer, index, MyBase.HeaderLength)
                EndianOrder.BigEndian.CopyBytes(IDCode, buffer, index)
                EndianOrder.BigEndian.CopyBytes(m_formatFlags, buffer, index + 2)
                EndianOrder.BigEndian.CopyBytes(Convert.ToInt16(PhasorDefinitions.Count), buffer, index + 4)
                EndianOrder.BigEndian.CopyBytes(Convert.ToInt16(AnalogDefinitions.Count), buffer, index + 6)
                EndianOrder.BigEndian.CopyBytes(Convert.ToInt16(DigitalDefinitions.Count), buffer, index + 8)

                Return buffer
            End Get
        End Property

        Protected Overrides Sub ParseHeaderImage(ByVal state As IChannelParsingState, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

            Dim parsingState As IConfigurationCellParsingState = DirectCast(state, IConfigurationCellParsingState)

            ' Parse out station name
            MyBase.ParseHeaderImage(state, binaryImage, startIndex)
            startIndex += MyBase.HeaderLength

            IDCode = EndianOrder.BigEndian.ToUInt16(binaryImage, startIndex)
            m_formatFlags = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 2)

            With parsingState
                .PhasorCount = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 4)
                .AnalogCount = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 6)
                .DigitalCount = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 8)
            End With

        End Sub

        Protected Overrides ReadOnly Property FooterLength() As UInt16
            Get
                Return MyBase.FooterLength + _
                    PhasorDefinitions.Count * PhasorDefinition.ConversionFactorLength + _
                    AnalogDefinitions.Count * AnalogDefinition.ConversionFactorLength + _
                    DigitalDefinitions.Count * DigitalDefinition.ConversionFactorLength + _
                    IIf(Parent.DraftRevision = DraftRevision.Draft7, 2, 0)
            End Get
        End Property

        Protected Overrides ReadOnly Property FooterImage() As Byte()
            Get
                Dim buffer As Byte() = CreateArray(Of Byte)(FooterLength)
                Dim x, index As Int32

                ' Include conversion factors in configuration cell footer
                With PhasorDefinitions
                    For x = 0 To .Count - 1
                        CopyImage(DirectCast(.Item(x), PhasorDefinition).ConversionFactorImage, buffer, index, PhasorDefinition.ConversionFactorLength)
                    Next
                End With

                With AnalogDefinitions
                    For x = 0 To .Count - 1
                        CopyImage(DirectCast(.Item(x), AnalogDefinition).ConversionFactorImage, buffer, index, AnalogDefinition.ConversionFactorLength)
                    Next
                End With

                With DigitalDefinitions
                    For x = 0 To .Count - 1
                        CopyImage(DirectCast(.Item(x), DigitalDefinition).ConversionFactorImage, buffer, index, DigitalDefinition.ConversionFactorLength)
                    Next
                End With

                ' Include nominal frequency
                CopyImage(MyBase.FooterImage, buffer, index, MyBase.FooterLength)

                ' Include configuration count (new for version 7.0)
                If Parent.DraftRevision = DraftRevision.Draft7 Then EndianOrder.BigEndian.CopyBytes(m_configurationCount, buffer, index)

                Return buffer
            End Get
        End Property

        Protected Overrides Sub ParseFooterImage(ByVal state As IChannelParsingState, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

            Dim x As Int32

            ' Parse conversion factors from configuration cell footer
            With PhasorDefinitions
                For x = 0 To .Count - 1
                    DirectCast(.Item(x), PhasorDefinition).ParseConversionFactor(binaryImage, startIndex)
                    startIndex += PhasorDefinition.ConversionFactorLength
                Next
            End With

            With AnalogDefinitions
                For x = 0 To .Count - 1
                    DirectCast(.Item(x), AnalogDefinition).ParseConversionFactor(binaryImage, startIndex)
                    startIndex += AnalogDefinition.ConversionFactorLength
                Next
            End With

            With DigitalDefinitions
                For x = 0 To .Count - 1
                    DirectCast(.Item(x), DigitalDefinition).ParseConversionFactor(binaryImage, startIndex)
                    startIndex += DigitalDefinition.ConversionFactorLength
                Next
            End With

            ' Parse nominal frequency
            MyBase.ParseFooterImage(state, binaryImage, startIndex)

            ' Get configuration count (new for version 7.0)
            If Parent.DraftRevision = DraftRevision.Draft7 Then m_configurationCount = EndianOrder.BigEndian.ToUInt16(binaryImage, startIndex)

        End Sub

        Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

            MyBase.GetObjectData(info, context)

            ' Serialize configuration cell
            info.AddValue("formatFlags", m_formatFlags, GetType(FormatFlags))
            info.AddValue("configurationCount", m_configurationCount)

        End Sub

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/IeeeC37_118/ConfigurationCell.cs
//*******************************************************************************************************
//  ConfigurationCell.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  11/12/2004 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.Serialization;
using System.Security.Permissions;

namespace PCS.PhasorProtocols.IeeeC37_118
{
    /// <summary>
    /// Represents the IEEE C37.118 implementation of a <see cref="IConfigurationCell"/> that can be sent or received.
    /// </summary>
    [Serializable()]
    public class ConfigurationCell : ConfigurationCellBase
    {
        #region [ Members ]

        // Fields
        private FormatFlags m_formatFlags;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="ConfigurationCell"/>.
        /// </summary>
        /// <param name="parent">The reference to parent <see cref="IConfigurationFrame"/> of this <see cref="ConfigurationCell"/>.</param>
        public ConfigurationCell(IConfigurationFrame parent)
            : base(parent, false, 0, Common.MaximumPhasorValues, Common.MaximumAnalogValues, Common.MaximumDigitalValues)
        {
            // Define new parsing state which defines constructors for key configuration values
            State = new ConfigurationCellParsingState(
                IeeeC37_118.PhasorDefinition.CreateNewDefinition,
                IeeeC37_118.FrequencyDefinition.CreateNewDefinition,
                IeeeC37_118.AnalogDefinition.CreateNewDefinition,
                IeeeC37_118.DigitalDefinition.CreateNewDefinition);
        }

        /// <summary>
        /// Creates a new <see cref="ConfigurationCell"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The reference to parent <see cref="ConfigurationFrame"/> of this <see cref="ConfigurationCell"/>.</param>
        /// <param name="idCode">The numeric ID code for this <see cref="ConfigurationCell"/>.</param>
        /// <param name="nominalFrequency">The nominal <see cref="LineFrequency"/> of the <see cref="FrequencyDefinition"/> of this <see cref="ConfigurationCell"/>.</param>
        public ConfigurationCell(ConfigurationFrame1 parent, ushort idCode, LineFrequency nominalFrequency)
            : this(parent)
        {
            IDCode = idCode;
            NominalFrequency = nominalFrequency;
        }

        /// <summary>
        /// Creates a new <see cref="ConfigurationCell"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected ConfigurationCell(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize configuration cell
            m_formatFlags = (FormatFlags)info.GetValue("formatFlags", typeof(FormatFlags));
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets a reference to the parent <see cref="ConfigurationFrame"/> for this <see cref="ConfigurationCell"/>.
        /// </summary>
        public new ConfigurationFrame1 Parent
        {
            get
            {
                return base.Parent as ConfigurationFrame1;
            }
            set
            {
                base.Parent = value;
            }
        }

        /// <summary>
        /// Gets or sets format flags of this <see cref="ConfigurationCell"/>.
        /// </summary>
        /// <remarks>
        /// These are bit flags, use properties to change basic values.
        /// </remarks>
        [EditorBrowsable(EditorBrowsableState.Advanced)]
        public FormatFlags FormatFlags
        {
            get
            {
                return m_formatFlags;
            }
            set
            {
                m_formatFlags = value;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="DataFormat"/> for the <see cref="IPhasorDefinition"/> objects in the <see cref="ConfigurationCellBase.PhasorDefinitions"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        public override DataFormat PhasorDataFormat
        {
            get
            {
                return (((m_formatFlags & FormatFlags.Phasors) > 0) ? DataFormat.FloatingPoint : DataFormat.FixedInteger);
            }
            set
            {
                if (value == DataFormat.FloatingPoint)
                    m_formatFlags = m_formatFlags | FormatFlags.Phasors;
                else
                    m_formatFlags = m_formatFlags & ~FormatFlags.Phasors;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="CoordinateFormat"/> for the <see cref="IPhasorDefinition"/> objects in the <see cref="ConfigurationCellBase.PhasorDefinitions"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        public override CoordinateFormat PhasorCoordinateFormat
        {
            get
            {
                return (((m_formatFlags & FormatFlags.Coordinates) > 0) ? CoordinateFormat.Polar : CoordinateFormat.Rectangular);
            }
            set
            {
                if (value == CoordinateFormat.Polar)
                    m_formatFlags = m_formatFlags | FormatFlags.Coordinates;
                else
                    m_formatFlags = m_formatFlags & ~FormatFlags.Coordinates;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="DataFormat"/> of the <see cref="FrequencyDefinition"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        public override DataFormat FrequencyDataFormat
        {
            get
            {
                return (((m_formatFlags & FormatFlags.Frequency) > 0) ? DataFormat.FloatingPoint : DataFormat.FixedInteger);
            }
            set
            {
                if (value == DataFormat.FloatingPoint)
                    m_formatFlags = m_formatFlags | FormatFlags.Frequency;
                else
                    m_formatFlags = m_formatFlags & ~FormatFlags.Frequency;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="DataFormat"/> for the <see cref="IAnalogDefinition"/> objects in the <see cref="ConfigurationCellBase.AnalogDefinitions"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        public override DataFormat AnalogDataFormat
        {
            get
            {
                return (((m_formatFlags & FormatFlags.Analog) > 0) ? DataFormat.FloatingPoint : DataFormat.FixedInteger);
            }
            set
            {
                if (value == DataFormat.FloatingPoint)
                    m_formatFlags = m_formatFlags | FormatFlags.Analog;
                else
                    m_formatFlags = m_formatFlags & ~FormatFlags.Analog;
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="HeaderImage"/>.
        /// </summary>
        protected override int HeaderLength
        {
            get
            {
                return base.HeaderLength + 10;
            }
        }

        /// <summary>
        /// Gets the binary header image of the <see cref="ConfigurationCell"/> object.
        /// </summary>
        protected override byte[] HeaderImage
        {
            get
            {
                byte[] buffer = new byte[HeaderLength];
                int index = 0;

                base.HeaderImage.CopyImage(buffer, ref index, base.HeaderLength);
                EndianOrder.BigEndian.CopyBytes(IDCode, buffer, index);
                EndianOrder.BigEndian.CopyBytes((ushort)m_formatFlags, buffer, index + 2);
                EndianOrder.BigEndian.CopyBytes((ushort)PhasorDefinitions.Count, buffer, index + 4);
                EndianOrder.BigEndian.CopyBytes((ushort)AnalogDefinitions.Count, buffer, index + 6);
                EndianOrder.BigEndian.CopyBytes((ushort)DigitalDefinitions.Count, buffer, index + 8);

                return buffer;
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="FooterImage"/>.
        /// </summary>
        protected override int FooterLength
        {
            get
            {
                return base.FooterLength + PhasorDefinitions.Count * PhasorDefinition.ConversionFactorLength + AnalogDefinitions.Count * AnalogDefinition.ConversionFactorLength + DigitalDefinitions.Count * DigitalDefinition.ConversionFactorLength + (Parent.DraftRevision > DraftRevision.Draft6 ? 2 : 0);
            }
        }

        /// <summary>
        /// Gets the binary footer image of the <see cref="ConfigurationCell"/> object.
        /// </summary>
        protected override byte[] FooterImage
        {
            get
            {
                byte[] buffer = new byte[FooterLength];
                PhasorDefinition phasorDefinition;
                AnalogDefinition analogDefinition;
                DigitalDefinition digitalDefinition;
                int x, index = 0;

                // Include conversion factors in configuration cell footer
                for (x = 0; x < PhasorDefinitions.Count; x++)
                {
                    phasorDefinition = PhasorDefinitions[x] as PhasorDefinition;

                    if (phasorDefinition != null)
                        phasorDefinition.ConversionFactorImage.CopyImage(buffer, ref index, PhasorDefinition.ConversionFactorLength);
                }

                for (x = 0; x < AnalogDefinitions.Count; x++)
                {
                    analogDefinition = AnalogDefinitions[x] as AnalogDefinition;

                    if (analogDefinition != null)
                        analogDefinition.ConversionFactorImage.CopyImage(buffer, ref index, AnalogDefinition.ConversionFactorLength);
                }

                for (x = 0; x < DigitalDefinitions.Count; x++)
                {
                    digitalDefinition = DigitalDefinitions[x] as DigitalDefinition;

                    if (digitalDefinition != null)
                        digitalDefinition.ConversionFactorImage.CopyImage(buffer, ref index, DigitalDefinition.ConversionFactorLength);
                }

                // Include nominal frequency
                base.FooterImage.CopyImage(buffer, ref index, base.FooterLength);

                // Include configuration count (new for version 7.0)
                if (Parent.DraftRevision > DraftRevision.Draft6)
                    EndianOrder.BigEndian.CopyBytes(RevisionCount, buffer, index);

                return buffer;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="ConfigurationCell"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Format Flags", (int)m_formatFlags + ": " + m_formatFlags);

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Parses the binary header image.
        /// </summary>
        /// <param name="binaryImage">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="binaryImage"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="binaryImage"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        protected override int ParseHeaderImage(byte[] binaryImage, int startIndex, int length)
        {
            IConfigurationCellParsingState state = State;
            int index = startIndex;

            // Parse out station name
            index += base.ParseHeaderImage(binaryImage, startIndex, length);

            IDCode = EndianOrder.BigEndian.ToUInt16(binaryImage, index);
            m_formatFlags = (FormatFlags)EndianOrder.BigEndian.ToUInt16(binaryImage, index + 2);

            // Parse out total phasors, analogs and digitals defined for this device
            state.PhasorCount = EndianOrder.BigEndian.ToUInt16(binaryImage, index + 4);
            state.AnalogCount = EndianOrder.BigEndian.ToUInt16(binaryImage, index + 6);
            state.DigitalCount = EndianOrder.BigEndian.ToUInt16(binaryImage, index + 8);
            index += 10;

            return (index - startIndex);
        }

        /// <summary>
        /// Parses the binary footer image.
        /// </summary>
        /// <param name="binaryImage">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="binaryImage"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="binaryImage"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        protected override int ParseFooterImage(byte[] binaryImage, int startIndex, int length)
        {
            PhasorDefinition phasorDefinition;
            AnalogDefinition analogDefinition;
            DigitalDefinition digitalDefinition;
            int x, index = startIndex;

            // Parse conversion factors from configuration cell footer
            for (x = 0; x < PhasorDefinitions.Count; x++)
            {
                phasorDefinition = PhasorDefinitions[x] as PhasorDefinition;

                if (phasorDefinition != null)
                    index += phasorDefinition.ParseConversionFactor(binaryImage, index);
            }

            for (x = 0; x < AnalogDefinitions.Count; x++)
            {
                analogDefinition = AnalogDefinitions[x] as AnalogDefinition;

                if (analogDefinition != null)
                    index += analogDefinition.ParseConversionFactor(binaryImage, index);
            }

            for (x = 0; x < DigitalDefinitions.Count; x++)
            {
                digitalDefinition = DigitalDefinitions[x] as DigitalDefinition;

                if (digitalDefinition != null)
                    index += digitalDefinition.ParseConversionFactor(binaryImage, index);
            }

            // Parse nominal frequency
            index += base.ParseFooterImage(binaryImage, index, length);

            // Parse out configuration count (new for version 7.0)
            if (Parent.DraftRevision > DraftRevision.Draft6)
            {
                RevisionCount = EndianOrder.BigEndian.ToUInt16(binaryImage, index);
                index += 2;
            }

            return (index - startIndex);
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize configuration cell
            info.AddValue("formatFlags", m_formatFlags, typeof(FormatFlags));
        }

        #endregion

        #region [ Static ]

        // Static Methods

        // Delegate handler to create a new IEEE C37.118 configuration cell
        internal static IConfigurationCell CreateNewCell(IChannelFrame parent, IChannelFrameParsingState<IConfigurationCell> state, int index, byte[] binaryImage, int startIndex, out int parsedLength)
        {
            ConfigurationCell configCell = new ConfigurationCell(parent as IConfigurationFrame);

            parsedLength = configCell.Initialize(binaryImage, startIndex, 0);

            return configCell;
        }

        #endregion
    }
}